src/__rt/pinned_stream.rs

```diff
@@ -0,0 +1,62 @@
+use std::ops::{Generator, GeneratorState};
+use std::marker::PhantomData;
+use std::mem;
+
+use anchor_experiment::{Pin, MovePinned};
+use futures::task;
+use futures::prelude::{Poll, Async};
+use stable::StableStream;
+
+use super::{CTX, Reset, IsResult};
+
+pub trait MyStableStream<T, U: IsResult<Ok=()>>: StableStream<Item=T, Error=U::Err> {}
+
+impl<F, T, U> MyStableStream<T, U> for F
+    where F: StableStream<Item = T, Error = U::Err> + ?Sized,
+          U: IsResult<Ok=()>
+{}
+
+/// Small shim to translate from a generator to a stream.
+struct GenStableStream<U, T> {
+    gen: T,
+    done: bool,
+    phantom: PhantomData<U>,
+}
+
+impl<U, T> !MovePinned for GenStableStream<U, T> { }
+
+pub fn gen_stream_pinned<T, U>(gen: T) -> impl MyStableStream<U, T::Return>
+    where T: Generator<Yield = Async<U>>,
+          T::Return: IsResult<Ok = ()>,
+{
+    GenStableStream { gen, done: false, phantom: PhantomData }
+}
+
+impl<U, T> StableStream for GenStableStream<U, T>
+    where T: Generator<Yield = Async<U>>,
+          T::Return: IsResult<Ok = ()>,
+{
+    type Item = U;
+    type Error = <T::Return as IsResult>::Err;
+
+    fn poll_next(mut self: Pin<Self>, ctx: &mut task::Context) -> Poll<Option<Self::Item>, Self::Error> {
+        CTX.with(|cell| {
```

Comment:
Ditto on this whole `transmute`-based repeated code-- you should be able to write a single bit of  `unsafe` code that would allow you to make this and other spots like it safe.